File: 20-actors/fuchi/methods/test_route.py
#!/usr/bin/env python3
"""Tests for 扶持 (fuchi) route.py — in-kind rail decomposition + governance gate.

Standalone-runnable: python3 test_route.py
"""
from __future__ import annotations

import sys

from route import (
    LINE_TO_RAIL,
    OPTIMISTIC_CEILING_USD_MICROS_YR,
    gov_route,
    in_kind_coverage,
    rider_hit,
    route_envelope,
    touches_invariant,
)


def _env(line, imputed, cash=0):
    return {":envelope/line": ":" + line, ":envelope/imputed-usd-micros-yr": imputed,
            ":envelope/cash-usd-micros": cash}


# ── route_envelope ──────────────────────────────────────────────────────────
def test_each_line_maps_to_its_producing_actor():
    rails = route_envelope([_env(l, 1_000_000) for l in
                            ("housing", "food", "energy", "compute", "tooling", "care", "liquidity")])
    kinds = {r.kind for r in rails}
    assert kinds == set(k for k, _ in LINE_TO_RAIL.values())
    providers = {r.provider_actor for r in rails}
    assert {"commons-land", "mitsuho", "hikari", "murakumo", "okaimono", "iyashi", "warifu"} == providers


def test_liquidity_is_member_principal_only():
    rails = route_envelope([_env("food", 1), _env("liquidity", 1)])
    by_kind = {r.kind: r for r in rails}
    assert by_kind["food-mitsuho"].member_principal is False
    assert by_kind["liquidity-warifu"].member_principal is True


def test_cash_line_is_unrepresentable():
    for bad in ("cash", "stipend", "cash-disbursement"):
        try:
            route_envelope([_env(bad, 1)])
        except ValueError as e:
            assert "cash≡0" in str(e)
            continue
        raise AssertionError(f"{bad!r} line must be refused (cash≡0)")


def test_nonzero_cash_field_is_refused():
    try:
        route_envelope([_env("food", 1, cash=5)])
    except ValueError as e:
        assert "cash≡0" in str(e)
        return
    raise AssertionError("nonzero :envelope/cash-usd-micros must be refused")


def test_unknown_line_has_no_rail():
    try:
        route_envelope([_env("yacht", 1)])
    except ValueError as e:
        assert "G3" in str(e)
        return
    raise AssertionError("an unknown line must have no rail (G3)")


# ── in_kind_coverage ────────────────────────────────────────────────────────
def test_fully_in_kind_is_100pct():
    rails = route_envelope([_env("food", 4), _env("energy", 1)])
    assert in_kind_coverage(rails) == 1.0


def test_liquidity_lowers_coverage():
    rails = route_envelope([_env("food", 5), _env("liquidity", 5)])
    assert in_kind_coverage(rails) == 0.5


# ── gov_route (G7 pure function) ────────────────────────────────────────────
def test_low_imputed_in_kind_auto_accepts():
    assert gov_route(1_000_000_000, invariant_touch=False, rider="") == "auto"


def test_above_ceiling_goes_to_vote():
    assert gov_route(OPTIMISTIC_CEILING_USD_MICROS_YR + 1, False, "") == "sbt-vote"


def test_invariant_touch_goes_to_council():
    assert gov_route(1, invariant_touch=True, rider="") == "council-lv7"


def test_rider_hit_is_refused_over_everything():
    # a rider hit beats both ceiling and invariant-touch → always refused
    assert gov_route(10 ** 18, invariant_touch=True, rider="affiliate") == "refused"


def test_rider_hit_detection():
    assert rider_hit("requests an affiliate ad-revenue share")
    assert rider_hit("広告 revenue")
    assert not rider_hit("maintains the sanae weeder")


def test_invariant_touch_detection():
    assert touches_invariant("new commons-land grant for housing")
    assert not touches_invariant("food and energy sustenance")


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"test_route.py: {len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/fuchi/methods/test_route.clj
;; test_route.clj — 扶持 route: in-kind rail decomposition + governance routing, parity with
;; route.py. Run via `bb test:fuchi`. ADR-2606142300.
(ns fuchi.methods.test-route
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [fuchi.methods.route :as r]))

(def ^:private envelope
  [{:envelope/line :housing   :envelope/imputed-usd-micros-yr 12000000000}
   {:envelope/line :food      :envelope/imputed-usd-micros-yr 6000000000}
   {:envelope/line :liquidity :envelope/imputed-usd-micros-yr 2000000000}])

(deftest envelope-decomposition
  (testing "envelope lines → in-kind rails over existing actors (golden from route.py)"
    (let [rails (r/route-envelope envelope)]
      (is (= [["housing-commons" "commons-land" false]
              ["food-mitsuho" "mitsuho" false]
              ["liquidity-warifu" "warifu" true]]                  ; liquidity = MEMBER-PRINCIPAL
             (mapv (juxt :kind :provider-actor :member-principal) rails)))
      (is (= [12000000000 6000000000 2000000000] (mapv :imputed-usd-micros-yr rails))))))

(deftest cash-unrepresentable
  (testing "cash≡0 — a cash/stipend rail or non-zero cash-micros RAISES; unknown line raises"
    (is (thrown? Exception (r/route-envelope [{:envelope/line :cash}])))
    (is (thrown? Exception (r/route-envelope [{:envelope/line :stipend}])))
    (is (thrown? Exception (r/route-envelope [{:envelope/line :housing :envelope/cash-usd-micros 5}])))
    (is (thrown? Exception (r/route-envelope [{:envelope/line :unknown-need}])))))

(deftest in-kind-coverage-metric
  (testing "fraction delivered in-kind (member-principal liquidity excluded), round 4"
    (is (== 0.9 (r/in-kind-coverage (r/route-envelope envelope))))   ; 18e9 / 20e9
    (is (== 1.0 (r/in-kind-coverage [])))                            ; empty → 1.0
    (is (== 1.0 (r/in-kind-coverage [{:imputed-usd-micros-yr 0 :member-principal false}])))))  ; total 0

(deftest governance-routing
  (testing "gov-route is a pure function of (total, invariant-touch, rider) — 非裁定"
    (is (= "auto"        (r/gov-route 1000 false "")))
    (is (= "sbt-vote"    (r/gov-route 25000000000 false "")))         ; above ceiling
    (is (= "council-lv7" (r/gov-route 1000 true "")))                 ; invariant touch
    (is (= "refused"     (r/gov-route 1000 false "advertis")))        ; rider hit
    (is (= "refused"     (r/gov-route 99999999999 true "weapon")))))  ; rider beats all

(deftest rider-and-invariant-scan
  (testing "rider-hit returns the matched token; touches-invariant flags constitutional contexts"
    (is (= "advertis" (r/rider-hit "buy ad space advertis here")))
    (is (= "weapon"   (r/rider-hit "this is a weapon system")))
    (is (= ""         (r/rider-hit "housing food energy")))
    (is (true?  (r/touches-invariant "new commons-land grant")))
    (is (true?  (r/touches-invariant "a charter amendment")))
    (is (false? (r/touches-invariant "food and energy")))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'fuchi.methods.test-route)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
